optuna/_deprecation.py

```diff
@@ -0,0 +1,92 @@
+import functools
+import inspect
+import textwrap
+from typing import Any
+from typing import Callable
+import warnings
+
+from optuna._experimental import _get_docstring_indent
+from optuna._experimental import _validate_version
+
+
+_DEPRECATION_NOTE_TEMPLATE = """
+
+.. note::
+    Deprecated in v{ver}. This feature will be removed in the future. See
+    https://github.com/optuna/optuna/releases/tag/v{ver}.
+"""
+
+
+def deprecation(version: str, name: str = None) -> Any:
+    """Decorate class or function as deprecation.
+
+    Args:
+        version: The first version that supports the target feature.
+        name: The name of the feature. Defaults to the function or class name. Optional.
+    """
+
+    _validate_version(version)
+
+    def _deprecation_wrapper(f: Any) -> Any:
+        # f is either func or class.
+
+        def _deprecation_func(func: Callable[[Any], Any]) -> Callable[[Any], Any]:
+            """Decorates a function as deprecated.
+
+            This decorator is supposed to be applied to the deprecated function.
+            """
+            if func.__doc__ is None:
+                func.__doc__ = ""
+
+            note = _DEPRECATION_NOTE_TEMPLATE.format(ver=version)
+            indent = _get_docstring_indent(func.__doc__)
+            func.__doc__ = func.__doc__.strip() + textwrap.indent(note, indent) + indent
+
+            # TODO(crcrpar): Annotate this correctly.
```

Comment:
A nit but probably best not to copy TODOs with names with developers. crcrpar is not necessarily aware/responsible for `_deprecation.py`. How about writing your own name?